### ISSUE TITLE
RFC: Arbitrary directory structures for packages 

### ISSUE THREAD

**Issue by threepointone** [jpmorganchase/modular#240]
This is a proposal for allowing any directory structure for packages/workspaces. 

(This is *not* about nested workspaces.) 

Currently, we hardcode all packages have to be `packages/*`. An example is the `modular` repository itself; all our packages are under the `packages` directory. 

This is fine for most setups; it's commonly used by plenty of open source projects and proprietary internal projects across companies. It breaks down when you have multiple teams working in the same repository tho. Common complaints I've heard are- "We'd like to have apps in one directory, and libs in another", "We'd like core stuff in one directory and contributed stuff in another", "Teams X and Y want different directory layouts", etc. I can sympathise with these opinions; there's no major technical reason for enforcing `packages/*` other than keeping stuff simple, and I still think that's a good default to have. 

Let's say we wanted to allow any possible directory structure, but without breaking any older repositories, how would we do it? 

First, we let the workspaces field in package.json be editable. It already is, but modular would break if you did it. Now let's say folks can do whatever they want to it. Cool, that also means this field is the source of truth for the value of `workspaces`. 

Let's also say we only allow packages anywhere under `/packages` (instead of just one level deep). (This leaves the root folder for meta information, and makes it a little easier implementation wise/backward compatibility. 

### `modular populate-workspaces`

A new command, that can scan an entire codebase, and automatically add any new workspaces to the workspaces field. This has 2 advantages for workflow. 

- In the simplest form, someone could make a new folder, add a package.json to it, and then run this command; if it didn't match any pattern in `workspaces`, it'll get added to it. 
- More interestingly, someone could take a whole glob of source code from elsewhere, potentially another modular codebase, dump it in another one, run the command, and get running with nothing else. 
- We can also use this command to verify that there aren't any clashing package names.

Maybe we could also make this run automatically anyone runs a plain `yarn` (by being a `prepare` hook? Dunno, must investigate.) This isn't critical, but nice to have.

### module boundaries 

This rule was always implied, but we never got around to enforcing it. With this change it becomes critical that we do so; "A workspace can only import from anywhere inside itself, or another workspace's main field (or `exports` field, which we'll use to specify extra entrypoints to the workspace.)"

### The `craco` config. 

The craco config exists to modify the webpack and jest configs to work with workspaces, this should be relatively trivial to read from the `workspaces` field. From this, it follows that 'apps' should just work, as would `modular test`. 

### `modular-views.macro`

This should also be fairly easy to modify to read matching directories from the `workspaces` field. 

### "Builds"

The build scripts will have to be changed to accommodate these changes, but while it isn't non-trivial, I don't expect it to be hard. BUT, a probable **breaking change** here, is that we can't refer to packages by their directory name anymore, we'll have to use their full package name (or directory name?). Assume modular scripts was called `@modular/scripts`. Instead of how we could currently say `modular build modular-scripts`, we'd have to say `modular build @modular/scripts` (or still `modular build modular-scripts` if the package's name is actually `modular-scripts`. )

### "Add" 

This is also technically a breaking change. We can't just say `modular add x`, we'll have to specify name and path explicitly. I'll leave this open for now so we can discuss what the devx for this should be like (maybe this is the time to pull the trigger on a gui?). In any case, you can also always make a folder with a package.json, and call `modular populate-packages`. 

### "Start" 

Similar to "Builds" and "Add", the change here is about how to reference packages. Again, maybe something to punt on to a gui. Or not! Happy to discuss. 

### Lint 

We don't have to change anything here, since the lint config applies to the entire project. 

### Typecheck

The one change we make here, is to change from 'packages/*' to 'packages/**'

--- 

There's probably some loose ends I've missed, I'll keep thinking about it. As you can see, it's not like we can do this without *some* breaking changes, but they're not so bad, and we're small enough right now that this is probably the right time to do it. This also gives us a good jumping off point for modular v2 (for which I'll write a different RFC later.)

**Comment by threepointone** [comment]
I'm working on a PR that addresses all of these.

**Comment by threepointone** [comment]
Question: What about enforcing package name styles? Like for `packages/xyz/*`, all packages must be named `@xyz/<something>`

Answer: This isn't something we should enforce in modular, but we don't block anyone writing a script that enforces this on their own codebase. (Implementation: run `yarn workspaces info` and validate the output)

**Comment by threepointone** [comment]
Thinking more about this (and working on the PR), `populate-packages` may not even be necessary. It's just a helper, sure, but people will probably also edit the field directly (and have terser patterns, etc). We'll see.

**Comment by threepointone** [comment]
I got most of this list done today, I now need to change the library build script, and add a bunch of tests to verify my changes. Not too bad, feeling confident about landing this later this week.

**Comment by threepointone** [comment]
I opened https://github.com/jpmorganchase/modular/pull/246 to address this RFC, have a read to see what I've implemented. Additionally: 
- It doesn't contain `populate-packages`. I implemented it, then realised that the output is too granular; when users would rather write their own wildcards anyway. Instead, new projects default to `packages/*` and users and free to narrow the scope down (or not!)
- It doesn't enforce module boundaries, that's something we should work on for `modular-check`.
